File: python/sorts/merge_sort.py
def merge_sort(a: list):
    n = len(a)

    # Base case
    if n <= 1:
        return

    # Divide list into 2 halves
    mid_index = n // 2
    left = a[:mid_index]
    right = a[mid_index:]

    # Sort 2 halves
    merge_sort(left)
    merge_sort(right)

    # From here, 2 halves are already sorted
    l = r = i = 0

    # Copy the higher number from each list (left to right) until all numbers of 1 list are copied
    while l < len(left) and r < len(right):
        current_left = left[l]
        current_right = right[r]
        if current_left < current_right:
            a[i] = current_left
            l += 1
        else:
            a[i] = current_right
            r += 1
        i += 1

    # Copy the remaining numbers from the other list
    while l < len(left):
        a[i] = left[l]
        l += 1
        i += 1

    while r < len(right):
        a[i] = right[r]
        r += 1
        i += 1

File: javascript/algorithms/sorts/merge_sort.js
/** Merge the two arrays: left and right
 * @param {Array} left - Left half of a root array
 * @param {Array} right - Right half of a root array
 */
function merge(left, right) {
	let resultArray = [],
		leftIndex = 0,
		rightIndex = 0;
	// We will concatenate values into the resultArray in order
	while (leftIndex < left.length && rightIndex < right.length) {
		if (left[leftIndex] < right[rightIndex]) {
			resultArray.push(left[leftIndex]);
			leftIndex++; // move left array cursor
		} else {
			resultArray.push(right[rightIndex]);
			rightIndex++; // move right array cursor
		}
	}

	// We need to concat here because there will be one element remaining
	// from either left OR the right
	return resultArray
		.concat(left.slice(leftIndex))
		.concat(right.slice(rightIndex));
}

/** Merge Sort Explanation
 * ------------------------------------------------------------
 * Description: Divide and Conquer Algorithm
 *
 * - Worst Case Time Complexity [ Big-O ]: O(n*log n)
 * - Best Case Time Complexity [Big-omega]: O(n*log n)
 * - Average Time Complexity [Big-theta]: O(n*log n)
 *
 * Explanation
 * - If r > l
 * - Find the middle point to divide the array into two halves:
 * -- middle m = (l+r)/2
 * - Call mergeSort for first half:
 * -- Call mergeSort(arr, l, m)
 * - Call mergeSort for second half:
 * -- Call mergeSort(arr, m+1, r)
 * - Merge the two halves sorted in step 2 and 3:
 * -- Call merge(arr, l, m, r)
 * ------------------------------------------------------------
 * @param {Array} arr - The root array
 * @returns {Array} The sorted array
 *
 * @chungquantin
 */
const mergeSort = function MergeSort(arr) {
	let len = arr.length;
	// Do not need to sort the array if it only has 1 element
	if (len <= 1) return arr;
	// Calculate the middle
	let m = Math.floor(len / 2);
	// Divide the left half
	let l = arr.slice(0, m);
	// Divide the right half
	let r = arr.slice(m);
	return merge(mergeSort(l), mergeSort(r));
};

module.exports = { mergeSort };
